src/object.hpp

```diff
@@ -0,0 +1,447 @@
+    [[nodiscard]] bool is_array() const noexcept { return type() == object_type::array; }
+    [[nodiscard]] bool is_string() const noexcept { return type() == object_type::string; }
+
+    [[nodiscard]] bool is_valid() const noexcept { return type() != object_type::invalid; }
+    [[nodiscard]] bool is_invalid() const noexcept { return type() == object_type::invalid; }
+
+    // Access the underlying value based on the required type
+    template <typename T>
+    [[nodiscard]] T as() const noexcept
+        requires std::is_same_v<T, Derived>
+    {
+        return *this;
+    }
+
+    template <typename T>
+    [[nodiscard]] T as() const noexcept
+        requires std::is_same_v<T, bool>
+    {
+        const auto &obj = static_cast<const Derived *>(this)->ref();
+        return obj.boolean;
+    }
+
+    template <typename T>
+    [[nodiscard]] T as() const noexcept
+        requires std::is_integral_v<T> && std::is_signed_v<T>
+    {
+        const auto &obj = static_cast<const Derived *>(this)->ref();
+        return static_cast<T>(obj.intValue);
+    }
+
+    template <typename T>
+    [[nodiscard]] T as() const noexcept
+        requires std::is_integral_v<T> && std::is_unsigned_v<T> && (!std::is_same_v<T, bool>)
+    {
+        const auto &obj = static_cast<const Derived *>(this)->ref();
+        return static_cast<T>(obj.uintValue);
+    }
+
+    template <typename T>
+    [[nodiscard]] T as() const noexcept
+        requires std::is_same_v<T, double>
+    {
+        const auto &obj = static_cast<const Derived *>(this)->ref();
+        return static_cast<T>(obj.f64);
+    }
+
+    template <typename T>
+    [[nodiscard]] T as() const noexcept
+        requires std::is_same_v<T, std::string> || std::is_same_v<T, std::string_view>
+    {
+        const auto &obj = static_cast<const Derived *>(this)->ref();
+        return {obj.stringValue, size()};
+    }
+
+    template <typename T>
+    [[nodiscard]] T as() const noexcept
+        requires std::is_same_v<T, const char *>
+    {
+        const auto &obj = static_cast<const Derived *>(this)->ref();
+        return obj.stringValue;
+    }
+
+    // Access the underlying value based on the required type or return a default
+    // value otherwise.
+    template <typename T> [[nodiscard]] T as_or_default(T default_value) const noexcept
+    {
+        const auto &obj = static_cast<const Derived *>(this)->ref();
+        if (!is_compatible_type<T>(static_cast<object_type>(obj.type))) {
+            [[unlikely]] return default_value;
+        }
+        return as<T>();
+    }
+
+private:
+    readable_object() = default;
+
+    friend Derived;
+};
+
+// TODO move object view and object converter header
+
+template <typename Derived> class writable_object {
+public:
+    [[nodiscard]] borrowed_object at(std::size_t idx);
+
+    borrowed_object emplace_back(owned_object &&value);
+    borrowed_object emplace(std::string_view key, owned_object &&value);
```

Comment:
These are also unsafe and I'd move them to separate types.

Reply:
Since everything is unsafe, there's no need to distinguish it...